Buffer H3Polygon2D bounding boxes to avoid edge precision issues

iverase opened the pull request:
We are using lucene spatial3d library to represent the area covered by a H3 bin. Unfortunately the mathematics at the edges are slightly different and the only way to determine if a point is inside or outside of a H3 bin is using the H3 library itself. In order to avoid issues at the edges when computing the spatial relationship of a bounding box with an H3 bin where are:

1) Computing the bounding box of the H3 Bin and expanding it by a factor to check for CELL_OUTSIDE_QUERY.
2) Expanding the inner nodes bounding box by a factor to check for CELL_INSIDE_QUERY.
3) Otherwise we will return  CELL_CROSSES_QUERY.

That should be enough to prevent border issues with a small performance penalty.

fixes https://github.com/elastic/elasticsearch/issues/87391
fixes https://github.com/elastic/elasticsearch/issues/89868

iverase committed:
Buffer H3Polygon2D bounding boxes to avoid edge precision issues

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

elasticsearchmachine commented:
Hi @iverase, I've created a changelog YAML for you.

iverase committed:
Update docs/changelog/89196.yaml

craigtaverner commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/query/H3LatLonGeometry.java:
```diff
@@ -117,18 +137,22 @@ public boolean contains(double x, double y) {
-                Math.toRadians(minY),
-                Math.toRadians(minX),
-                Math.toRadians(maxX)
+                Math.toRadians(Math.min(GeoUtils.MAX_LAT_INCL, maxY + BBOX_BUFFER_FRACTION)),
```
When you calculated the min/max above you used `BBOX_BUFFER_FRACTION * (maxX - minX)`, but here you just use buffer_fraction alone. I think it makes more sense to multiply by the box width, or by the hexagon width.

iverase replied:
Actually, I just want to increase the edge surface by a fix amount regardless of the size of the bounding box. I have renamed  BBOX_BUFFER_FRACTION to BBOX_EDGE_DELTA and just add a fix value to the edges.

craigtaverner commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/query/H3LatLonGeometry.java:
```diff
@@ -117,18 +137,22 @@ public boolean contains(double x, double y) {
-                Math.toRadians(minY),
-                Math.toRadians(minX),
-                Math.toRadians(maxX)
+                Math.toRadians(Math.min(GeoUtils.MAX_LAT_INCL, maxY + BBOX_BUFFER_FRACTION)),
```
```suggestion
                Math.toRadians(Math.min(GeoUtils.MAX_LAT_INCL, maxY + BBOX_BUFFER_FRACTION * (maxY - minY))),
```

craigtaverner commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/query/H3LatLonGeometry.java:
```diff
@@ -117,18 +137,22 @@ public boolean contains(double x, double y) {
-                Math.toRadians(minX),
-                Math.toRadians(maxX)
+                Math.toRadians(Math.min(GeoUtils.MAX_LAT_INCL, maxY + BBOX_BUFFER_FRACTION)),
+                Math.toRadians(Math.max(GeoUtils.MIN_LAT_INCL, minY - BBOX_BUFFER_FRACTION)),
```
```suggestion
                Math.toRadians(Math.max(GeoUtils.MIN_LAT_INCL, minY - BBOX_BUFFER_FRACTION * (maxY - minY))),
```

craigtaverner commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/query/H3LatLonGeometry.java:
```diff
@@ -117,18 +137,22 @@ public boolean contains(double x, double y) {
-                Math.toRadians(maxX)
+                Math.toRadians(Math.min(GeoUtils.MAX_LAT_INCL, maxY + BBOX_BUFFER_FRACTION)),
+                Math.toRadians(Math.max(GeoUtils.MIN_LAT_INCL, minY - BBOX_BUFFER_FRACTION)),
+                Math.toRadians(Math.max(GeoUtils.MIN_LON_INCL, minX - BBOX_BUFFER_FRACTION)),
```
```suggestion
                Math.toRadians(Math.max(GeoUtils.MIN_LON_INCL, minX - BBOX_BUFFER_FRACTION * (maxX - minX))),
```

craigtaverner commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/query/H3LatLonGeometry.java:
```diff
@@ -117,18 +137,22 @@ public boolean contains(double x, double y) {
+                Math.toRadians(Math.min(GeoUtils.MAX_LAT_INCL, maxY + BBOX_BUFFER_FRACTION)),
+                Math.toRadians(Math.max(GeoUtils.MIN_LAT_INCL, minY - BBOX_BUFFER_FRACTION)),
+                Math.toRadians(Math.max(GeoUtils.MIN_LON_INCL, minX - BBOX_BUFFER_FRACTION)),
+                Math.toRadians(Math.min(GeoUtils.MAX_LON_INCL, maxX + BBOX_BUFFER_FRACTION))
```
```suggestion
                Math.toRadians(Math.min(GeoUtils.MAX_LON_INCL, maxX + BBOX_BUFFER_FRACTION * (maxX - minX)))
```

craigtaverner commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/query/H3LatLonGeometry.java:
```diff
@@ -117,18 +137,22 @@ public boolean contains(double x, double y) {
-                case GeoArea.DISJOINT -> PointValues.Relation.CELL_OUTSIDE_QUERY;
-                default -> PointValues.Relation.CELL_CROSSES_QUERY;
-            };
+            return box.getRelationship(hexagon) == GeoArea.CONTAINS
```
Is there not a case here where it can be outside, but within the buffer?

iverase replied:
oh yes

craigtaverner reviewed (changes requested):
I liked this, but did think you needed to use the BUFFER_FRACTION multiplied by the width or height of the cell or search box.

craigtaverner commented:
Also, it feels like this code should have unit tests to test explicitly for the inside/outside/crosses cases.

iverase committed:
Merge branch 'main' into fix87391

iverase committed:
Merge branch 'fix87391' of github.com:iverase/elasticsearch into fix87391

iverase committed:
address review comments

iverase commented:
I am looking into which unit test might make sense here.

craigtaverner reviewed (approved):
Looks much better. Would still be nice with some unit tests, but I like this a lot.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into fix87391

craigtaverner committed:
Add more tests for H3LatLonGeometry.relate method

These tests are simplified versions of tests written for the extended
H3LatLonGeometry class that supports all Component2D interface methods.
That work was for supporting H3 aggregations over geo_shape.

craigtaverner commented:
I pushed some more tests that test for INSIDE/OUTSIDE/CROSSING for points just inside and just outside the H3 cell, but we see test failures quite often with the randomized test, especially for high latitudes. I think the current `relate` method is not quite working.

elasticsearchmachine commented:
Hi @iverase, I've updated the changelog YAML for you.

iverase committed:
Update docs/changelog/89196.yaml

craigtaverner commented:
Also, should this fix remove the `AwaitsFix` annotations on the two tests that have it?

iverase committed:
Merge branch 'main' into fix87391

iverase committed:
Merge branch 'fix87391' of github.com:iverase/elasticsearch into fix87391

iverase committed:
iter
